File: ruby/testcode/pathtraver_033.rb
require_relative 'shared'

# vuln-code-snippet start ruby_pt_realpath_prefix_check
def read_realpath_prefix(req)
  base = '/app/files'
  joined = File.join(base, req.param('rel'))
  rp = File.realpath(joined) rescue nil
  raise 'traversal' unless rp&.start_with?(base + '/')
  File.read(rp) # vuln-code-snippet safe-line ruby_pt_realpath_prefix_check
  BenchmarkResponse.ok(File.read(rp))
end
# vuln-code-snippet end ruby_pt_realpath_prefix_check

File: php/testcode/pathtraver_033.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_pt_pathinfo_ext_prefix
function pathtraver033(BenchmarkRequest $req): BenchmarkResponse {
    $f = $req->param('file');
    $ext = pathinfo($f, PATHINFO_EXTENSION);
    $allowed_ext = ['jpg', 'png', 'gif'];
    if (!in_array(strtolower($ext), $allowed_ext, true)) {
        return BenchmarkResponse::badRequest('invalid extension');
    }
    $content = file_get_contents('/var/app/uploads/' . basename($f)); // vuln-code-snippet safe-line php_pt_pathinfo_ext_prefix
    return BenchmarkResponse::ok($content);
}
// vuln-code-snippet end php_pt_pathinfo_ext_prefix
